client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -239,6 +240,12 @@ boolean isWithinRefreshThreshold(Instant currentInstant, Instant expiration) {
                     this.tokenRefreshFuture.compareAndSet(handleToCurrentRefreshFuture, null);
                     LoggerHelpers.traceLeave(log, "refreshToken", traceEnterId, this.scopeName, this.streamName);
                     return delegationToken.get().getValue();
+                })
+                .exceptionally(ex -> {
```

Comment:
we an use `whenComplete((token, throwable) -> {}` if `this.recreateToken()` returns a CompletableFuture<DelegationToken>. This would simplify the code right ?

Reply:
The [JavaDocs ](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html) for CompletableFuture says that `whenComplete` _"returns a new CompletionStage with the same result or exception as this stage, that executes the given action when this stage completes"_.

In this piece of code, we are not returning the result of the previous CompletionStage (except for the exception itself), but are rather returning an object extracted inside this stage: `delegationToken.get().getValue()`. 

Therefore, I believe that `whenComplete` is not a good choice here.

Reply:
I think if we modify `this.recreateToken()` returns a CompletableFuture<String> or CompletableFuture<DelegationToken> then this would simply the code here.

Reply:
As discussed, I have modified the code to remove duplication via use of the `handle()` method.